Change user id and group id to 999 for postgres user

rcmadhankumar opened the pull request.

dirkmueller commented on src/bci_build/package/postgres.py:
```diff
@@ -82,7 +82,24 @@
+{DOCKERFILE_RUN} groupmod -g 999 postgres; \
+    usermod -u 999 postgres; \
+    chown -R postgres:postgres /var/lib/pgsql; \
+    chmod +x /usr/local/bin/docker-entrypoint.sh; \
```
I think what we should be doing is create a base container layer that includes those users, then you don't need to do `chown -R` and preserve the permissions set by installing the packages.

rcmadhankumar replied:
approach changed.

dirkmueller commented on src/bci_build/package/postgres.py:
```diff
@@ -82,7 +82,24 @@
+    usermod -u 999 postgres; \
+    chown -R postgres:postgres /var/lib/pgsql; \
+    chmod +x /usr/local/bin/docker-entrypoint.sh; \
+    sed -i -e 's/exec gosu postgres "/exec setpriv --reuid=postgres --regid=postgres --clear-groups -- "/g' /usr/local/bin/docker-entrypoint.sh; \
```
this is unrelated isn't it?

alexandrevicenzi commented on src/bci_build/package/postgres.py:
```diff
@@ -82,7 +82,24 @@
+HEALTHCHECK --interval=10s --start-period=10s --timeout=5s \
+    CMD pg_isready -U ${{POSTGRES_USER:-postgres}} -h localhost -p 5432
+"""
+            if ver == 16
```
`ver` is the psql version,  we should use `os_version` and apply this to SLE 16 only, to any version on SLE 16.

alexandrevicenzi replied:
perhaps also apply to Tumbleweed, but ideally not to SLE 15.

rcmadhankumar replied:
The stable user and group ids will be supported only from postgres version 18, made changes accordingly.
Added postgres version 17 to the list just for testing right now.

rcmadhankumar commented:
looking into failures. out for review for the approach and steps.

rcmadhankumar commented on src/bci_build/templates.py:
```diff
@@ -53,7 +53,18 @@
+    OLD_UID=$(chroot /target id -u {{ image.stable.user_name }});\\
+    chroot /target groupmod -g {{ image.stable.group_id }} {{ image.stable.group_name }}; \\
+    chroot /target usermod -u {{ image.stable.user_id }} {{ image.stable.user_name }}; \\
+    chroot /target  find / \\( -gid $OLD_GID -o -uid $OLD_UID \\) -exec chown {{ image.stable.user_id }}:{{ image.stable.group_id }} {} \\;
```
tested and fixed.

rcmadhankumar commented on src/bci_build/templates.py:
```diff
@@ -53,7 +53,18 @@
+{%- endif %}
+{% if image.stable %}# changing user id and group id created by package installation to stable values
+{{ DOCKERFILE_RUN }} \\
+    OLD_GID=$(chroot /target getent group {{ image.stable.group_name }}| cut -d: -f3);\\
```
added a condition and make changes only if the user and group exist.

rcmadhankumar commented on src/bci_build/package/__init__.py:
```diff
@@ -77,6 +77,22 @@ class ParseVersion(enum.StrEnum):
 
 
+@dataclass
+class Stable:
```
changed.

rcmadhankumar commented on src/bci_build/package/postgres.py:
```diff
@@ -20,6 +21,9 @@
 # first list the SLE15 versions, then the TW specific versions
 _POSTGRES_MAJOR_VERSIONS = [17, 16, 15, 14] + [13]
+# postgres versions that supports stable user id and stable group ids
+# added 17 just for testing, will be supported only from 18
```
this will be shipped with 18. however postgres 18 is not available right now.
Which means for the demo we cannot see the dockerfile from github ci incase if I had added version as 17 here.

Once the PR is approved, then i will change this value to 18 and we can merge this PR. the change will stay and will never get executed until we release postgres 18 image.

rcmadhankumar replied:
made changes to the comment and changed to expected value.
PR is ready for merge.

dirkmueller reviewed (changes requested):
Please add migration instructions to the README.md that explain what is happening?

dirkmueller commented on src/bci_build/templates.py:
```diff
@@ -53,7 +53,16 @@
+{%- endif %}
+{% if image.stable_user %}# changing user id and group id created by package installation to stable values
+{{ DOCKERFILE_RUN }} \\
+    if {% if image.from_target_image %}chroot /target {% endif %}getent passwd {{ image.stable_user.user_name }} && {% if image.from_target_image %}chroot /target {% endif %}getent group {{ image.stable_user.group_name }}; then\\
```
why are we checking for the user to exist? IMHO this should hard fail otherwise we do not find mismatches at container build time - which is dangerous imho.

rcmadhankumar replied:
removed

rcmadhankumar commented on src/bci_build/package/postgres.py:
```diff
@@ -84,7 +88,16 @@
         volumes=["$PGDATA"],
         exposes_ports=[TCP(5432)],
-        custom_end=rf"""COPY docker-entrypoint.sh /usr/local/bin/
+        stable_user=StableUser(
```
added.

rcmadhankumar commented on src/bci_build/package/__init__.py:
```diff
@@ -194,6 +210,9 @@ class BaseContainerImage(abc.ABC):
     entrypoint: list[str] | None = None
 
+    # user&group that needs to be stabilized for the package
+    stable_user: StableUser | None = None
```
changed.

rcmadhankumar commented:
> Please add migration instructions to the README.md that explain what is happening?

Added a readme section that explains the user and group id changes in postgres version 18.
(Since, we are not supporting migration on postgress yet, I am assuming user is going to take the backup as sql file and execute it and added info considering the same use case)
